[LOOP-1315] Upload Alert history to Tidepool backend

darinkrauss opened the pull request:
- https://tidepool.atlassian.net/browse/LOOP-1315
- Add alert data support to RemoteDataService
- Add parameters property to Alert
- Add SyncAlertObject

Commit:
[LOOP-1315] Upload Alert history to Tidepool backend

- https://tidepool.atlassian.net/browse/LOOP-1315
- Add alert data support to RemoteDataService
- Add parameters property to Alert
- Add SyncAlertObject

darinkrauss commented:
LW PR for build: https://github.com/tidepool-org/LoopWorkspace/pull/757

rickpasetto commented on LoopKit/Alert.swift:
```diff
@@ -94,15 +94,19 @@ public struct Alert: Equatable {
     public let sound: Sound?
+
+    /// Any parameters used to dynamically construct foreground or background content encoded as a JSON object string
+    public let parameters: String?
```
I think I'm still not convinced that the best way to store this here at this level is a JSON-encoded (i.e. post-serialized) string.  I think my preference still would be as a Dictionary, (i.e. `[String: Any]`).  Feel free to make a `typealias` if it helps.

darinkrauss replied:
The problem here is that decoding the resultant JSON string stored in Core Data won't necessarily decode back into the expected data types. For example (and just as an example, not that you'd do this), if you stored an `Alert.Trigger` in the dictionary, then after it did a round-trip from `Alert` to Core Data (via `StoredAlert`) back to `Alert` what once was an `Alert.Trigger` is now a `String` (because `Alert.Trigger` encodes as a `String` in JSON). The only way I see to make this consistent and not error prone would be to create a brand new type of dictionary (like `UserDefaults) that **only** accepts the core JSON data types. That seems like way too much unnecessary complication for me.

rickpasetto replied:
But that doesn't need to be enforced, does it?  Can't we just declare that, by convention, the Values in this dictionary have to be valid JSON (or plist, whichever you prefer) types?  This is not that uncommon, in my experience.

darinkrauss replied:
That would not be my preference. If everyone else feels otherwise, I'll change it.

rickpasetto replied:
It just seems wrong to me to put a serialized string in here as part of the LoopKit API.  I'm fine moving forward if I'm the only one who objects, but I still do object.

ps2 replied:
It's already a fairly accepted convention for some interfaces, such as UserDefaults, to require simple types (basically plist types), so I think it could be documented as such and we wouldn't need to invent a new type system. It does seem a little more natural to me that users of this api abide by that rule, rather than have to deal with json encoding/decoding for a "metadata" field, which a simple dictionary should work great for.

darinkrauss replied:
Yea, I'm going to need more details here for both of your suggestions. I'll add it to the list of things to discuss at dev sync.

rickpasetto reviewed (commented):
See my comment.  It sounded like you may have been open to using a Dictionary instead of a String.  I think that would be my preference.

darinkrauss commented:
@rickpasetto Responded to your comment. Actually, I wasn't really open to using a Dictionary because of the problems I describe in my response, but I said that I'd let it go to PR/code review and we'd discuss it from here.

Commit:
Updates based upon PR feedback

- Rename alert parameters to metadata
- Use dictionary instead of string for alert metadata
- Add AnyCodableEquatable

darinkrauss commented:
Updated to use dictionary rather that string to capture metadata (was parameters).

Commit:
Fix decoding issue with AnyCodableEquatable

rickpasetto commented on LoopKit/AnyCodableEquatable.swift:
```diff
@@ -0,0 +1,46 @@
+        case unknownType
+    }
+
+    public let any: Any
```
[nit] A variable named `any`, while correct, I guess, seems a little strange to me.  Maybe consider just `value`?  Another common name is `wrapped`.  Up to you.

darinkrauss replied:
Changed to `wrapped`.

rickpasetto commented on LoopKit/AnyCodableEquatable.swift:
```diff
@@ -0,0 +1,46 @@
+
+    public init(from decoder: Decoder) throws {
+        let container = try decoder.singleValueContainer()
+        if let value = try? container.decode(String.self) {
```
I thought you needed `Date`...did you not want to use that?

darinkrauss replied:
Turns out I was storing a TimeInterval and not a Date so I removed this. If we need it later we can add it back.

darinkrauss replied:
Essentially, I just went with the basic JSON data types (less object and array).

rickpasetto commented on LoopKit/AnyCodableEquatable.swift:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    public init(from decoder: Decoder) throws {
+        let container = try decoder.singleValueContainer()
```
Out of curiosity, why this and not `unkeyedContainer`?

darinkrauss replied:
An `unkeyedContainer` is used for arrays while `singleValueContainer` is used for primitive values.

rickpasetto reviewed (approved):
Approving because I don't want to block progress, but conditional on two things:
1. Didn't you want to also support `Date` as a type in the `Metadata`?  If not, it is fine as-is, but I was just wondering if maybe it was left out by accident
2. Please write `Codable` and `Equatable` unit tests for `AnyCodableEquatable`.  Thanks.

rickpasetto commented on LoopKit/Alert.swift:
```diff
@@ -94,15 +94,21 @@ public struct Alert: Equatable {
+        self.metadata = metadata
     }
 }
 
```
Just to make things a little more ergonomic, consider adding an extension so Metadata construction is a little cleaner:

```swift
extension Alert.Metadata {
    init<E: Codable & Equatable>(_ dict: [String: E]) {
        self = dict.mapValues { AnyCodableEquatable($0) }
    }
}
```
That way you can initialize `Metadata` like this:
```swift
let metadata = Alert.Metadata(["Foo": "Bar"])
```

Just a suggestion...

darinkrauss replied:
This will work for simple cases (dictionary values are all of the same type), but not for more complex cases (mixed value type dictionary). However, added and used where possible.

Commit:
Updates based upon PR feedback

ps2 reviewed (approved):
LGTM!

Commit:
Merge branch 'dev' into darinkrauss/LOOP-1315/upload-alerts
